transports/quic/src/hole_punching.rs

```diff
@@ -33,10 +32,8 @@ async fn punch_holes<P: Provider>(socket: UdpSocket, remote_addr: SocketAddr) ->
             .take(64)
             .collect();
 
-        if let Err(e) = socket.send_to(&contents, remote_addr) {
-            if !matches!(e.kind(), io::ErrorKind::WouldBlock) {
-                return Error::Io(e);
-            }
+        if let Err(e) = P::send_to(&socket, &contents, remote_addr).await {
+            return Error::Io(e);
         }
```

Comment:
Isn't this just what `?` would do?

Reply:
I wondered the same. Note the function signature:

``` rust
async fn punch_holes<P: Provider>(socket: UdpSocket, remote_addr: SocketAddr) -> Error {
```

The function never returns, unless there is an `Error`. In words, the function returns `Error` and not `Result<_, Error>`. Thus one can not use `?`.

Reply:
Perhaps we should change it to `Result<Infallible, Error>`?

Reply:
We discussed this in https://github.com/libp2p/rust-libp2p/pull/3964/ as well, can't find the concrete conversation right now. I suggest we keep as is unless you feel strongly about it @thomaseizinger.